docs/getting_started/developer-guide.md

```diff
@@ -0,0 +1,248 @@
+should do that first. If you want to see a more general user guide, look
+at the [oras.land Python
+guide](https://oras.land/client_libraries/1_python/).
+
+## Documentation
+
+The documentation is provided in the `docs` folder of the repository,
+and generally most content that you might want to add is under
+`getting_started`. For ease of contribution, files that are likely to be
+updated by contributors (e.g., mostly everything but the module generated files)
+ are written in markdown. If you need to use [toctree](https://www.sphinx-doc.org/en/master/usage/restructuredtext/directives.html#table-of-contents) you should not use extra newlines or spaces (see index.md files for exampls).
+Markdown is the chosen language for the oras community, and this is why we chose to
+use it over restructured syntax - it makes it easier to contribute documentation.
+
+
+### Install Dependencies and Build
+
+The documentation is built using sphinx, and generally you can install
+dependencies:
+
+``` console
+# in main oras-py folder (oras-py is needed to build docs)
+$ pip install -e .
+
+# Now docs dependencies
+cd docs
+pip install -r requrements.txt
+
+# Build the docs into _build/html
+make html
+```
+
+### Preview Documentation
+
+After `make html` you can cd into `_build/html` and start a local web
+server to preview:
+
+``` console
+$ python -m http.server 9999
+```
+
+And open your browser to `localhost:9999`
+
+### Docstrings
+
+To render our Python API into the docs, we keep an updated restructured
+syntax in the `docs/source` folder that you can update on demand as
+follows:
+
+``` console
+$ ./apidoc.sh
+```
+
+This should only be required if you change any docstrings or add/remove
+functions from oras-py source code.
+
+## Creating a Custom Client
+
+If you want to create your own client, you likely want to:
+
+1.  Subclass the `oras.provider.Registry` class
+2.  Add custom functions to push, pull, or create manifests / layers
+3.  Provide authentication, if required.
+
+### The Registry Client
+
+Your most basic registry (that will mimic the default can be created
+like this:
+
+``` python
+import oras.provider
+
+class MyProvider(oras.provider.Registry):
+    pass
+```
+
+### Pull Interactions
+
+You can imagine having a custom function that will retrieve a manifest
+or blob and do custom operations on it.
+
+``` python
+def inspect(self, name):
+
+    # Parse the name into a container
+    container = self.get_container(name)
+
+    # Get the manifest with the three layers
+    manifest = self.get_manifest(container)
+
+    # Organize layers based on media_types
+    layers = self._organize_layers(manifest)
+
+    # Read info.json media type, find correct blob...
+    # download correct blob, etc.
+```
+
+Specifically the function `self.get_blob` will allow you to do that.
+
+### Push Interactions
+
+You might instead want to have a custom lookup of archive paths and
+media types. Let\'s say we start with this lookup, `archives`:
```

Comment:
Do we need to remove `\` in `Let\'s`?

Reply:
Yes I think so!

Reply:
okay I think I got most of them - let me know if you see one I missed! https://github.com/oras-project/oras-py/pull/41/commits/4a66955cbe84ebb0db6d030772f07e374f722574